File: backend/app/schemas/holdings.py
"""
TradeNova Backend — Holdings Schemas
"""

from pydantic import BaseModel
from typing import List


class HoldingItem(BaseModel):
    """Individual holding"""
    tradingsymbol: str
    exchange: str
    isin: str
    company_name: str
    quantity: int
    t1_quantity: int
    average_price: float
    last_price: float
    close_price: float
    pnl: float
    pnl_percentage: float
    day_change: float
    day_change_percentage: float
    current_value: float
    invested_value: float


class PortfolioSummary(BaseModel):
    """Aggregated portfolio metrics"""
    total_investment: float
    current_value: float
    total_pnl: float
    total_pnl_percentage: float
    day_pnl: float
    day_pnl_percentage: float
    total_holdings: int


class HoldingsResponse(BaseModel):
    """Full holdings response with summary"""
    holdings: List[HoldingItem]
    summary: PortfolioSummary

File: tradenova/types/holdings.ts
/**
 * Holdings type definitions
 * Maps to Upstox Holdings API response structure
 */

/** Individual holding item */
export interface Holding {
  tradingsymbol: string;
  exchange: string;
  isin: string;
  quantity: number;
  t1_quantity: number;
  average_price: number;
  last_price: number;
  pnl: number;
  day_change: number;
  day_change_percentage: number;
  close_price: number;
  /** Computed fields */
  current_value: number;
  invested_value: number;
  pnl_percentage: number;
  /** Company display name */
  company_name: string;
}

/** Portfolio summary computed from holdings */
export interface PortfolioSummary {
  total_investment: number;
  current_value: number;
  total_pnl: number;
  total_pnl_percentage: number;
  day_pnl: number;
  day_pnl_percentage: number;
  total_holdings: number;
}

/** Full holdings API response */
export interface HoldingsResponse {
  holdings: Holding[];
  summary: PortfolioSummary;
}
